cpp/src/quantiles/groupby.cu

```diff
@@ -0,0 +1,264 @@
+template <bool nullable = true>
+struct transform_row_eq_comparator {
+  row_equality_comparator<nullable> cmp;
+  gdf_size_type const* sorted_order;
+
+  CUDA_DEVICE_CALLABLE
+  bool operator() (gdf_size_type lhs, gdf_size_type rhs) {
+    return cmp(sorted_order[lhs], sorted_order[rhs]);
+  }
+};
+
+struct permutation_label_setter {
+  gdf_size_type * group_labels_ptr;
+  gdf_size_type const* group_ids_ptr;
+
+  CUDA_DEVICE_CALLABLE
+  void operator() (gdf_size_type i) { 
+    group_labels_ptr[group_ids_ptr[i]] = 1;
+  }
+};
+
+} // namespace anonymous
+
+
+namespace cudf {
+
+namespace detail {
+
+gdf_column const& groupby::key_sort_order() {
+  if (_key_sorted_order)
+    return *_key_sorted_order;
+
+  _key_sorted_order = std::make_unique<gdf_column>(
+    allocate_column(gdf_dtype_of<gdf_index_type>(),
+                    _key_table.num_rows(),
+                    false));
+
+  if (_include_nulls ||
+      !cudf::has_nulls(_key_table)) {  // SQL style
+    gdf_context context{};
+    context.flag_groupby_include_nulls = true;
+    CUDF_TRY(gdf_order_by(_key_table.begin(), nullptr,
+                          _key_table.num_columns(), _key_sorted_order.get(),
+                          &context));
+  } else {  // Pandas style
+    auto key_cols_bitmask = row_bitmask(_key_table);
+
+    gdf_column modified_fist_key_col = *(_key_table.get_column(0));
+    modified_fist_key_col.valid =
+        reinterpret_cast<gdf_valid_type*>(key_cols_bitmask.data().get());
+
+    auto keys = const_cast<cudf::table*>(&_key_table);
+    std::vector<gdf_column*> modified_key_cols_vect(keys->begin(), keys->end());
+    modified_key_cols_vect[0] = &modified_fist_key_col;
+    cudf::table modified_key_col_table(modified_key_cols_vect.data(),
+                                      modified_key_cols_vect.size());
+
+    gdf_context temp_ctx;
+    temp_ctx.flag_null_sort_behavior = GDF_NULL_AS_LARGEST;
+
+    CUDF_TRY(gdf_order_by(modified_key_col_table.begin(), nullptr,
+                          modified_key_col_table.num_columns(),
+                          _key_sorted_order.get(), &temp_ctx));
+
+    CUDF_TRY(gdf_count_nonzero_mask(
+        reinterpret_cast<gdf_valid_type*>(key_cols_bitmask.data().get()),
+        _key_table.num_rows(),
+        &_num_keys));
+  }
+
+  return *_key_sorted_order;
+}
+
+rmm::device_vector<gdf_size_type> const& groupby::group_indices() {
+  if (_group_ids)
+    return *_group_ids;
+
+  index_vector idx_data(_num_keys);
+
+  auto counting_iter = thrust::make_counting_iterator<gdf_size_type>(0);
+  auto device_input_table = device_table::create(_key_table);
+  bool nullable = device_input_table.get()->has_nulls();
+  auto sorted_order = reinterpret_cast<gdf_size_type*>(key_sort_order().data);
+  decltype(idx_data.begin()) result_end;
+
+  if (nullable) {
+    auto comp = row_equality_comparator<true>(*device_input_table, true);
+    result_end = thrust::unique_copy(
+      thrust::device, counting_iter, counting_iter + _num_keys,
```

Comment:
Use rmm exec policy, not `thrust::device` -- unique_copy probably allocates temporary memory.